diff --git a/Txl/norm.Txl b/Txl/norm.Txl
index 23094e1f2e0d115156ec4273b12c77b3c61b9e66..29cf805365486216ce14b9c83dde64bc886273d7 100644
--- a/Txl/norm.Txl
+++ b/Txl/norm.Txl
@@ -95,11 +95,8 @@ function update_id I [typeid]
      'end 'define
      'function find_replace_I 
 	    'replace '[ 'program '] 'P '[ 'program ']
-	    'import 'TXLargs '[ 'repeat 'stringlit '] 
-	    'deconstruct '* 'TXLargs
-		"-diff" 'Filename '[ 'stringlit '] 
-	    'construct 'P2 '[ 'program '] _ '[ 'read 'Filename '] '[ 'program_normalise ']
-	    'construct 'S '[ TypeID '* '] '_ '[ find_clones_I 'P 'P2 ']
+	    'import 'Program_Diff '['program']
+	    'construct 'S '[ TypeID '* '] '_ '[ find_clones_I 'P 'Program_Diff '] 
 	    'construct 'P3 '[ 'program '] 'P '[ mark_clones_I 'each 'S ']
 	    'by 'P3 
      'end 'function
@@ -112,15 +109,17 @@ function update_id I [typeid]
 	    'construct 'S2 '[ TypeID '* '] '_ '[ '^ 'P2 ']
 	    'deconstruct '* 'S2 'Stmt 'Rest '[ TypeID '* '] 
 	    'replace '[ TypeID '* '] 'StructuredClones '[ TypeID '* ']
-	    'deconstruct 'not '* 'StructuredClones '<<< % 'Stmt
+	    'deconstruct 'not '* 'StructuredClones 'C '['attr 'Change '] % 'Stmt
 	    'by 'StructuredClones '[ '. 'Stmt '] 
      'end 'function
      'function mark_clones_I 'S '[ TypeID ']
 	    'import 'CloneNumber '['number']
+	    'import 'Program_Diff '['program']
 	    'export 'CloneNumber 'CloneNumber '['+ '1']
 	    'construct 'S1 '[ TypeID '] '<<< % 'S1 
-	    'replace '[ 'program '] 'P '[ 'program '] 'by
-		'P '[ '$ 'S 'S1 ']
+	    'construct 'S2 '[ TypeID '] '>>> % 'S1 
+	    'export 'Program_Diff 'Program_Diff  '[ '$ 'S 'S2 ']
+	    'replace '[ 'program '] 'P '[ 'program '] 'by 'P '[ '$ 'S 'S1 ']
      'end 'function
  export Rules Rules [. S1] 
  export KeepRuleIDs KeepRuleIDs [. find_replace_I] 
@@ -607,18 +606,38 @@ function main
 	'end 'function
  	'function 'program_remove_clone 'replace '[ 'program '] 'Prg '[ 'program '] 
     	  'export 'CloneNumber '['number'] '0
+	  'import 'TXLargs '[ 'repeat 'stringlit '] 
+	  'deconstruct '* 'TXLargs
+		"-diff" 'Filename '[ 'stringlit '] 
+	  'construct 'P2 '[ 'program '] _ '[ 'read 'Filename '] '[ 'program_normalise ']
+	  'export 'Program_Diff '['program'] 'P2
 	  'by R_RemoveClone
 	'end 'function
-        'redefine 'program
-	   '... '| '[ empty ']
-	'end 'define
+	'redefine 'program '... '| '[ 'empty '] '| '[ 'attr 'number '] 'end 'define
+	'function 'print_diff_1
+		'replace '[ 'program '] 'Prg '[ 'program ']
+		'construct 'P2 '['program'] 'Prg '[ 'write "t.t" '] '[ 'read "t.t" '] 
+		'construct 'S_Diff '['stringlit'] '_ '['quote 'P2 ']
+		'where 'not 'S_Diff '['= "" '] 
+		'by 'Prg
+	'end 'function
+	'function 'print_diff_2
+		'replace '[ 'program '] 'Prg '[ 'program ']
+		'construct 'P2 '['program'] 'Prg '[ 'write "t.t" '] '[ 'read "t.t" '] 
+		'construct 'S_Diff '['stringlit'] '_ '['quote 'P2 ']
+		'where 'S_Diff '['= "" '] 
+		'by '_
+	'end 'function
 	'function 'print_diff
 		'replace '[ 'program '] 'P_diff '[ 'program ']
 		'import 'CloneNumber '[ 'number ']
 		'where 'CloneNumber '[ '> '0 ']
-		'construct 'N_Clone '[ 'number '] 'CloneNumber '[ 'print ']
-		'construct 'P2 '[ 'program '] 'P_diff '[ 'printattr ']
-		'by _
+		'import 'Program_Diff '[ 'program ']
+		'% output the differences
+		'construct 'P_Clone '[ 'program '] 'CloneNumber
+		'construct 'P_Left '['program'] 'P_diff '[print_diff_1'] '[print_diff_2'] '['printattr']
+		'construct 'P_Right '['program'] 'Program_Diff '[print_diff_1'] '[print_diff_2'] '['printattr']
+		'by 'P_Clone
 	'end 'function
 	'function 'main 'replace '[ 'program '] 'Prg '[ 'program '] 
 	  'by 'Prg '[ 'program_normalise ']
